### PR TITLE
Login endpoint

### PR DESCRIPTION
**Pull request by mickamcia**
this is just backend, endpoint is usable

### CODE REVIEW

File: .vs/11/v17/.wsuo
Comment (KarolDenst) [inline .vs/11/v17/.wsuo:1]: what is this?

File: backend/YouTubeV2.Application/Services/JwtFeatures/JwtSettings.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        internal double ExpiresInDays { get; }
+
+        public JwtSettings(IConfiguration configurationSection)
+        {
+            SecurityKey = Encoding.UTF8.GetBytes(configurationSection["securityKey"]);
+            ValidIssuer = configurationSection["validIssuer"];
+            ValidAudience = configurationSection["validAudience"];
+            ExpiresInDays = Convert.ToDouble(configurationSection["expiresInDays"]);
+        }
+        public JwtSettings()
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Services/JwtFeatures/JwtSettings.cs]: DO we use this?

File: backend/YouTubeV2.Application/Services/UserService.cs [hunk 2]
```diff
@@ -15,12 +17,33 @@ public class UserService
...
+            _jwtHandler = jwtHandler;
+        }
+
+        public async Task<LoginResponseDto> LoginAsync(LoginDto loginDto, CancellationToken cancellationToken)
+        {
+            await _loginDtoValidator.ValidateAndThrowAsync(loginDto, cancellationToken);
+
+            User user = await GetUserByEmailAsync(loginDto.email);
+
+            if (!await _userManager.IsEmailConfirmedAsync(user))
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Services/UserService.cs]: we don't confirm emails. Its not in the specification

File: backend/YouTubeV2.Application/Services/UserService.cs [hunk 3]
```diff
@@ -15,12 +17,33 @@ public class UserService
...
+                throw new BadRequestException(new ErrorResponseDTO[1] { new ErrorResponseDTO("Email is not confirmed") });
+
+            if (!await _userManager.CheckPasswordAsync(user, loginDto.password))
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Services/UserService.cs:39]: i don't think you need '1' in brackets. [1] -> []

File: backend/YouTubeV2.Application/Services/UserService.cs [hunk 4]
```diff
@@ -15,12 +17,33 @@ public class UserService
...
+                throw new BadRequestException(new ErrorResponseDTO[1] { new ErrorResponseDTO("Provided password is invalid") });
+
+            return new LoginResponseDto(await _jwtHandler.GenerateTokenAsync(user));
+            //return new LoginResponseDto("dupa");
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Services/UserService.cs]: I don't think this comment add much

File: backend/YouTubeV2.Application/Services/UserService.cs [hunk 5]
```diff
@@ -53,5 +76,12 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
             byte[] image = Convert.FromBase64String(registerDto.avatarImage);
             await _blobImageService.UploadProfilePictureAsync(image, user.Id, cancellationToken);
         }
+        private async Task<User> GetUserByEmailAsync(string email)
+        {
+            var user = await _userManager.FindByEmailAsync(email);
+            if (user == null)
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Services/UserService.cs]: I'd move this to the validator

File: backend/YouTubeV2.Application/Validator/LoginDtoValidator.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+using YouTubeV2.Application.DTO;
+using YouTubeV2.Application.Model;
+
+namespace YouTubeV2.Application.Validator
+{
+    public class LoginDtoValidator : AbstractValidator<LoginDto>
+    {
+        public LoginDtoValidator(UserManager<User> userManager)
+        {
+            RuleFor(x => x.email)
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/Validator/LoginDtoValidator.cs:13]: You can add more validation. Like if email is a valida email or maybe check if a user with this email exists

File: backend/YouTubeV2.Application/YouTubeV2.Application.csproj [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (KarolDenst) [inline backend/YouTubeV2.Application/YouTubeV2.Application.csproj:1]: What is the change here?

File: backend/YouTubeV2/Controllers/UserController.cs [hunk 8]
```diff
@@ -21,5 +21,11 @@ public async Task<IActionResult> RegisterAsync([FromBody] RegisterDto registerDt
 
             return Ok();
         }
+        [HttpPost("login")]
+        public async Task<IActionResult> LoginAsync(LoginDto loginDto, CancellationToken cancellationToken)
+        {
+            LoginResponseDto loginResponseDto = await _userService.LoginAsync(loginDto, cancellationToken);
+            return Ok(loginResponseDto);
```
Comment (KarolDenst) [inline backend/YouTubeV2/Controllers/UserController.cs:30]: add empty line before return

File: backend/YouTubeV2/Program.cs [hunk 9]
```diff
@@ -29,6 +33,7 @@
 builder.Services.AddSingleton<IBlobImageService, BlobImageService>();
 
 builder.Services.AddValidatorsFromAssemblyContaining<RegisterDtoValidator>();
+builder.Services.AddValidatorsFromAssemblyContaining<LoginDtoValidator>();
```
Comment (KarolDenst) [inline backend/YouTubeV2/Program.cs]: this line does nothing. AddValidatorsFromAssemblyContaining adds all validators that are in the same folder as registerDtoValidator so if LoginDtoValidator is in the same folder we're just doing the same thing a secont time

File: backend/YouTubeV2/Program.cs [hunk 10]
```diff
@@ -44,6 +49,33 @@
             .WithExposedHeaders("Content-Disposition"));
 });
 
+builder.Services.Configure<DataProtectionTokenProviderOptions>(options =>
+       options.TokenLifespan = TimeSpan.FromHours(2));
+
+var jwtSettings = new JwtSettings(builder.Configuration.GetSection("JWTSettings"));
+//var jwtSettings = new JwtSettings();
```
Comment (KarolDenst) [inline backend/YouTubeV2/Program.cs]: remove commented code

File: backend/YouTubeV2/Program.cs [hunk 11]
```diff
@@ -44,6 +49,33 @@
...
+builder.Services.AddSingleton(jwtSettings);
+
+builder.Services.AddAuthentication(options =>
+{
+    options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
+    options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+})
+.AddJwtBearer(options =>
```
Comment (KarolDenst) [inline backend/YouTubeV2/Program.cs]: add indent

Comment (KarolDenst) [reply backend/YouTubeV2/Program.cs]: and maybe move addauthentication to the next line

File: backend/YouTubeV2/Program.cs [hunk 12]
```diff
@@ -44,6 +49,33 @@
...
+        ValidateIssuer = true,
+        ValidateAudience = true,
+        ValidateLifetime = true,
+        ValidateIssuerSigningKey = true,
+        ValidIssuer = jwtSettings.ValidIssuer,
+        ValidAudience = jwtSettings.ValidAudience,
+        IssuerSigningKey = new SymmetricSecurityKey(jwtSettings.SecurityKey)
+    };
+});
+builder.Services.AddScoped<JwtHandler>();
```
Comment (KarolDenst) [inline backend/YouTubeV2/Program.cs]: move this so that all services are added in the same place. Maybe even create a function that adds all services

Comment (KarolDenst) [reply backend/YouTubeV2/Program.cs]: also why not transient?

**Comment by KarolDenst** [comment]
image

**Comment by KarolDenst** [comment]
Did you check if loggin works? Maybe create a test endpoint to show that we can get the user from the token we return in /login

File: backend/YouTubeV2.Application/Services/UserService.cs (hunk 2 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/Services/UserService.cs]: deleted

File: backend/YouTubeV2.Application/Services/UserService.cs (hunk 3 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/Services/UserService.cs:39]: deleted

File: backend/YouTubeV2.Application/Services/UserService.cs (hunk 4 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/Services/UserService.cs]: deleted

File: backend/YouTubeV2.Application/Services/UserService.cs (hunk 5 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/Services/UserService.cs]: moved

File: backend/YouTubeV2.Application/Validator/LoginDtoValidator.cs (hunk 6 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/Validator/LoginDtoValidator.cs:13]: added

File: backend/YouTubeV2.Application/YouTubeV2.Application.csproj (hunk 7 above)
Comment (mickamcia) [reply backend/YouTubeV2.Application/YouTubeV2.Application.csproj:1]: it is a non-change

File: backend/YouTubeV2/Controllers/UserController.cs (hunk 8 above)
Comment (mickamcia) [reply backend/YouTubeV2/Controllers/UserController.cs:30]: added

File: backend/YouTubeV2/Program.cs (hunk 9 above)
Comment (mickamcia) [reply backend/YouTubeV2/Program.cs]: deleted

File: backend/YouTubeV2/Program.cs (hunk 10 above)
Comment (mickamcia) [reply backend/YouTubeV2/Program.cs]: removed

File: backend/YouTubeV2/Program.cs (hunk 11 above)
Comment (mickamcia) [reply backend/YouTubeV2/Program.cs]: added

File: backend/YouTubeV2/Program.cs (hunk 12 above)
Comment (mickamcia) [reply backend/YouTubeV2/Program.cs]: moved and changed to transient

File: .vs/11/v17/.wsuo
Comment (mickamcia) [reply .vs/11/v17/.wsuo:1]: deleted local file

**Comment by mickamcia** [comment]
regarding token that is being returned in /login, i only check if it is nonempty,

File: backend/YouTubeV2.Api.Tests/UserTests.cs [hunk 13]
```diff
@@ -63,5 +65,28 @@ public async Task RegisterShouldAddToDB()
...
         }
+        [TestMethod]
+        public async Task LoginUserShouldReturnOk()
+        {
+            // ARRANGE
+            var httpClient = _webApplicationFactory.CreateClient();
+
+            var registerDto = new RegisterDto("mail@mail.com", "Senior", "Generator", "Frajdy", "asdf1243@#$GJH", Role.Simple, "");
+
+            _ = await httpClient.PostAsync("register", new StringContent(JsonConvert.SerializeObject(registerDto), Encoding.UTF8, "application/json"));
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Api.Tests/UserTests.cs]: dont use endpoint just add user using usermanager

File: backend/YouTubeV2/Program.cs [hunk 14]
```diff
@@ -1,68 +1,102 @@
...
+
+        app.UseHttpsRedirection();
+
+        app.UseAuthentication();
+        app.UseAuthorization();
+
+        app.MapControllers();
+
+        app.Run();
+    }
```
Comment (Marvvvin5) [inline backend/YouTubeV2/Program.cs:45]: new line

File: backend/YouTubeV2.Api.Tests/UserTests.cs (hunk 13 above)
Comment (mickamcia) [reply backend/YouTubeV2.Api.Tests/UserTests.cs]: modified

File: backend/YouTubeV2.Api.Tests/UserTests.cs [hunk 15]
```diff
@@ -87,6 +93,7 @@ public async Task LoginUserShouldReturnOk()
             loginResponseDto.token.Should().NotBeNullOrEmpty();
             response.Content.Should().NotBeNull();
             response.StatusCode.Should().Be(HttpStatusCode.OK);
+
```
Comment (adaszkiewicza) [inline backend/YouTubeV2.Api.Tests/UserTests.cs]: remove empty line

File: backend/YouTubeV2/Program.cs [hunk 16]
```diff
@@ -41,6 +41,7 @@ public static void Main(string[] args)
         app.MapControllers();
 
         app.Run();
+
```
Comment (adaszkiewicza) [inline backend/YouTubeV2/Program.cs]: empty line

**Review by KarolDenst — APPROVED** [review]

**Pull request merged**